writer/writer.go

```diff
@@ -15,27 +15,380 @@
+	"type-info":       {Path: "type-info", IsDir: false, Required: true},
+	"checksums":       {Path: "checksums", IsDir: true, Required: false},
+	"checksums/*":     {Path: "checksums", IsDir: false, Required: false},
+	"signatures":      {Path: "signatures", IsDir: true, Required: true},
+	"signatures/*":    {Path: "signatures", IsDir: false, Required: true},
+	"scripts":         {Path: "scripts", IsDir: true, Required: false},
+	"scripts/pre":     {Path: "scripts/pre", IsDir: true, Required: false},
+	"scripts/pre/*":   {Path: "scripts/pre", IsDir: false, Required: false},
+	"scripts/post":    {Path: "scripts/post", IsDir: true, Required: false},
+	"scripts/post/*":  {Path: "scripts/post", IsDir: false, Required: false},
+	"scripts/check":   {Path: "scripts/check", IsDir: true, Required: false},
+	"scripts/check/*": {Path: "scripts/check/*", IsDir: false, Required: false},
+	// we must have data directory containing update
+	"data":   {Path: "data", IsDir: true, Required: true},
+	"data/*": {Path: "data/*", IsDir: false, Required: true},
+}
+
+// NewArtifactsWriter creates a new ArtifactsWriter providing a location
+// of Mender metadata artifacts, format of the update and version.
+func NewArtifactsWriter(path string, format string, version int) *ArtifactsWriter {
+	return &ArtifactsWriter{
+		updateLocation:  path,
+		headerStructure: metadata.MetadataArtifactHeader{Artifacts: ArtifactsHeaderFormat},
+		format:          format,
+		version:         version,
+		updates:         make(map[string]updateBucket),
+	}
+}
+
+type updateArtifact struct {
+	name         string
+	path         string
+	updateBucket string
+	info         os.FileInfo
+	checksum     []byte
+}
+
+type updateBucket struct {
+	location        string
+	path            string
+	archivedPath    string
+	updateArtifacts []updateArtifact
+	files           metadata.MetadataFiles
+}
+
+// ReadArchiver provides interface for reading files or streams and preparing
+// those to be written to tar archive.
+// GetHeader returns Header to be written to the crrent entry it the tar archive.
+type ReadArchiver interface {
+	io.ReadCloser
+	Open() error
+	GetHeader() (*tar.Header, error)
+}
+
+func (av ArtifactsWriter) calculateChecksum(upd *updateArtifact) error {
+	f, err := os.Open(upd.path)
+	if err != nil {
+		return errors.Wrapf(err, "can not open file for calculating checksum")
+	}
+	defer f.Close()
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return errors.Wrapf(err, "error calculating checksum")
+	}
+
+	checksum := h.Sum(nil)
+	upd.checksum = make([]byte, hex.EncodedLen(len(checksum)))
+	hex.Encode(upd.checksum, h.Sum(nil))
+	log.Debugf("hash of file: %v (%x)\n", upd.path, upd.checksum)
+	return nil
+}
+
+func (av ArtifactsWriter) getHeaderInfo(updates []os.FileInfo) metadata.MetadataHeaderInfo {
+	// for now we have ONLY one type of update - rootfs-image
+	headerInfo := metadata.MetadataHeaderInfo{}
+
+	// TODO: should we store update name as well?
```

Comment:
No, there is a specific reason I avoided this: An update is not guaranteed to have just one payload file, and some might not have any at all. Some updates may provide a URL instead, or login details somewhere, and this will be in the `type-info` header. In order to keep the format modular, when reading the `header-info` file you should not need to know anything about the expected contents of an update, only what type it is. This allows you to skip or ignore update types you don't know about.

We could combine `files` and `type-info` (they were combined in the first draft actually). The reason I split them is to (hopefully) make the logic simpler across different update type implementations: `files` will usually be the same format in most update types, hence can share code, whereas `type-info` is very likely to be different. Not sure if this is a great reason, but at least the content of `files` should be inside the `headers` folder, not in the `header-info` header itself.